ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamTestUtils.java

```diff
@@ -343,10 +343,8 @@ static void assertHeader(RaftServer server, RaftClientRequest header, int dataSi
 
   static void assertRaftClientMessage(RaftClientMessage expected, RaftPeerId expectedServerId, RaftClientMessage computed) {
     Assert.assertNotNull(computed);
-    Assert.assertEquals(expected.getClientId(), computed.getClientId());
```

Comment:
> In more details, Primary receives a RaftClientRequest R from a client as usual. Then, it encodes R as a Message, puts it in a new RaftClientRequest F and sends it to the Leader. Since Leader sees that the Type is ForwarededRequestTypeProto, it will decode the Message back to R and then processes R.

RaftClientRequest R -> primary's client wrap RaftClientRequest R in RaftClientRequest F -> leader decode R from F -> leader response RaftClientReply of R -> leader convert RaftClientReply of R to RaftClientReply of F.

Because the ClientId and CallId of F's RaftClientReply belongs to primary's client. So we can not check the ClientId and CallId of F's  RaftClientReply.

Reply:
Similar to expectedServerId, let's pass the clientId in Primary as the expectedClientId (a new parameter).

Reply:
@szetszwo Thanks the suggestions. I have updated the patch.
As you find, if we expose primary's ClientId to user, the test looks complicated. Besides, maybe we should not expose  primary's ClientId to user,  it maybe confuse user, because user will find RaftClientReply's ClientId does not match his RaftClient's ClientId. So maybe primary can convert RaftClientReply of F to RaftClientReply of R again. what do you think ?

Reply:
In the test, we may pass the primary server in the builder.
```
+++ b/ratis-test/src/test/java/org/apache/ratis/datastream/DataStreamAsyncClusterTests.java
@@ -106,8 +106,7 @@ public abstract class DataStreamAsyncClusterTests<CLUSTER extends MiniRaftCluste
         .orElseThrow(IllegalStateException::new);
   }
 
-  ClientId getPrimaryClientId(CLUSTER cluster, RaftClient client) {
-    RaftPeer primary = client.getPrimaryDataStreamServer();
+  ClientId getPrimaryClientId(CLUSTER cluster, RaftPeer primary) {
     return cluster.getDivision(primary.getId()).getRaftClient().getId();
   }
 
@@ -115,8 +114,9 @@ public abstract class DataStreamAsyncClusterTests<CLUSTER extends MiniRaftCluste
     final Iterable<RaftServer> servers = CollectionUtils.as(cluster.getServers(), s -> s);
     final RaftPeerId leader = cluster.getLeader().getId();
     final List<CompletableFuture<RaftClientReply>> futures = new ArrayList<>();
-    try(RaftClient client = cluster.createClient()) {
-      ClientId primaryClientId = getPrimaryClientId(cluster, client);
+    final RaftPeer primaryServer = CollectionUtils.random(cluster.getGroup().getPeers());
+    try(RaftClient client = cluster.createClient(primaryServer)) {
+      ClientId primaryClientId = getPrimaryClientId(cluster, primaryServer);
       for (int i = 0; i < numStreams; i++) {
         final DataStreamOutputImpl out = (DataStreamOutputImpl) client.getDataStreamApi().stream();
         futures.add(CompletableFuture.supplyAsync(() -> DataStreamTestUtils.writeAndCloseAndAssertReplies(
```